File: src/vb/Samples/Samples/Samples/SolidEdge/Application/ReportProperties.vb
Imports ApiSamples.Samples.SolidEdge
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace ApiSamples.Samples.SolidEdge
	''' <summary>
	''' Reports the file properties of the active document.
	''' </summary>
	Friend Class ReportProperties
		Private Shared Sub RunSample(ByVal breakOnStart As Boolean)
			If breakOnStart Then
				System.Diagnostics.Debugger.Break()
			End If

			Dim application As SolidEdgeFramework.Application = Nothing
			Dim document As SolidEdgeFramework.SolidEdgeDocument = Nothing
			Dim propertySets As SolidEdgeFramework.PropertySets = Nothing
			Dim properties As SolidEdgeFramework.Properties = Nothing
			Dim [property] As SolidEdgeFramework.Property = Nothing

			Try
				' Register with OLE to handle concurrency issues on the current thread.
				OleMessageFilter.Register()

				' Connect to or start Solid Edge.
				application = ApplicationHelper.Connect(True)

				' Make sure user can see the GUI.
				application.Visible = True

				' Bring Solid Edge to the foreground.
				application.Activate()

				' Get a reference to the active document.
				document = application.TryActiveDocumentAs(Of SolidEdgeFramework.SolidEdgeDocument)()

				If document IsNot Nothing Then
					' Get a reference to the Properties collection.
					propertySets = DirectCast(document.Properties, SolidEdgeFramework.PropertySets)

					For i As Integer = 1 To propertySets.Count
						properties = propertySets.Item(i)

						Console.WriteLine("PropertSet '{0}'.", properties.Name)

						For j As Integer = 1 To properties.Count
							Dim nativePropertyType As System.Runtime.InteropServices.VarEnum = System.Runtime.InteropServices.VarEnum.VT_EMPTY
							Dim runtimePropertyType As Type = Nothing

							Dim value As Object = Nothing

							[property] = properties.Item(j)
							nativePropertyType = CType([property].Type, System.Runtime.InteropServices.VarEnum)

							' Accessing Value property may throw an exception...
							Try
								value = [property].get_Value()
							Catch ex As System.Exception
								value = ex.Message
							End Try

							If value IsNot Nothing Then
								runtimePropertyType = value.GetType()
							End If

							Console.WriteLine(ControlChars.Tab & "{0} = '{1}' ({2} | {3}).", [property].Name, value, nativePropertyType, runtimePropertyType)
						Next j

						Console.WriteLine()
					Next i
				Else
					Throw New System.Exception(Resources.NoActiveDocument)
				End If

			Catch ex As System.Exception
				Console.WriteLine(ex.Message)
			Finally
				OleMessageFilter.Unregister()
			End Try
		End Sub
	End Class
End Namespace

File: src/cs/Samples/Samples/Samples/SolidEdge/Application/ReportProperties.cs
using SolidEdgeCommunity; //SolidEdge.Community.dll
using SolidEdgeFramework.Extensions; //SolidEdge.Community.dll
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ApiSamples.Samples.SolidEdge
{
    /// <summary>
    /// Reports the file properties of the active document.
    /// </summary>
    class ReportProperties
    {
        static void RunSample(bool breakOnStart)
        {
            if (breakOnStart) System.Diagnostics.Debugger.Break();

            SolidEdgeFramework.Application application = null;
            SolidEdgeFramework.SolidEdgeDocument document = null;
            SolidEdgeFramework.PropertySets propertySets = null;
            SolidEdgeFramework.Properties properties = null;
            SolidEdgeFramework.Property property = null;

            try
            {
                // Register with OLE to handle concurrency issues on the current thread.
                OleMessageFilter.Register();

                // Connect to or start Solid Edge.
                application = SolidEdgeCommunity.SolidEdgeInstall.Connect(true);

                // Make sure user can see the GUI.
                application.Visible = true;

                // Bring Solid Edge to the foreground.
                application.Activate();

                // Get a reference to the active document.
                document = application.GetActiveDocument<SolidEdgeFramework.SolidEdgeDocument>(false);

                if (document != null)
                {
                    // Get a reference to the Properties collection.
                    propertySets = (SolidEdgeFramework.PropertySets)document.Properties;

                    for (int i = 1; i <= propertySets.Count; i++)
                    {
                        properties = propertySets.Item(i);

                        Console.WriteLine("PropertSet '{0}'.", properties.Name);

                        for (int j = 1; j <= properties.Count; j++)
                        {
                            System.Runtime.InteropServices.VarEnum nativePropertyType = System.Runtime.InteropServices.VarEnum.VT_EMPTY;
                            Type runtimePropertyType = null;

                            object value = null;

                            property = properties.Item(j);
                            nativePropertyType = (System.Runtime.InteropServices.VarEnum)property.Type;

                            // Accessing Value property may throw an exception...
                            try
                            {
                                value = property.get_Value();
                            }
                            catch (System.Exception ex)
                            {
                                value = ex.Message;
                            }

                            if (value != null)
                            {
                                runtimePropertyType = value.GetType();
                            }

                            Console.WriteLine("\t{0} = '{1}' ({2} | {3}).", property.Name, value, nativePropertyType, runtimePropertyType);
                        }

                        Console.WriteLine();
                    }
                }
                else
                {
                    throw new System.Exception(Resources.NoActiveDocument);
                }

            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                OleMessageFilter.Unregister();
            }
        }
    }
}
